### ISSUE TITLE
Graph Predictor Module

### ISSUE THREAD

**Issue by Tomorrowdawn** [Tomorrowdawn/EvolveGCN#2]
## Component

这是一个专门为本次实验设计的预测器

以下内容已经过期. 需要重构.

### 预测任务

给定一组社交网络和提案信息, 预测每个成员对于提案的态度: YES, NO 或者 Not Vote

## Excepted Input and output

Input: glist, a list of DGLGraph. Each graph contains two attributes: ndata['feat'] and edata['weight'].

A batched proposal dict
```
proposal: 一个字典，包含'id','sponsors'和'cosponsors'。
        分别代表提案id(一个[B, 1] tensor)，发起者（一个[B, 1] tensor），协助发起者(一个[B, 4] tensor. 因为我们发现95%分位点就是4个协助发起者)
        如果不足4个协力, 则用-1填充。
        返回一个[B, N, 3]张量，表示每个节点yes/no/none的概率。
```

Output: A prediction tensor, shape of [B, N, 3]. Each representing yes/no/not vote possibility.

## Mechanics

该Predictor由两个类构成, 分别是RepresentationModule和ClassifierModule.

RepresentationModule是一个EvolveGCN网络, 拆掉了分类头, 输出一个[N, H]的节点embedding.

ClassifierModule是一个注意力汇聚网络, 接收任意多个sponsor embeddings和待预测的成员的embedding, 生成一个[1, 3] tensor. 

### 测试项目

从训练集中留出10%(相对于初始数据集)数据, 在训练后进行检测, 

检测指标: Accuracy, Loss

**Comment by Tomorrowdawn** [comment]
commit 35f1695 已完成. 主要类为class ProposalPredictor(nn.Module):

TODO: sampled_members. 如果训练量过大, 使用采样.

**Comment by Tomorrowdawn** [comment]
get_embeddings()待测试: 下标为-1的indice取出的是否确实为全零向量.

是否符合预期逻辑: 对于给定形如[B, S] 的indices, 输出为[B, S, V], 每一行和indices一一对应(result[i][j] = embedding[indices[i][j]]).

**Comment by Tomorrowdawn** [comment]
重构: 

现有代码的计算性能过差. 每次都会计算整个window并取最后一个window的hidden states.

在训练时, 不需要每次都只计算一个固定Window下的, 而是分成两步:

```py
for each timestep:
     current graph = get_graph(timestep)
     hidden states = model(timestep)
     for batch in proposals: 
           predict = predictor(hiddenstates, batch)
           compute loss and backward.
```

这样可以避免hidden states重复计算的开销. model内部维护一个随着time step变化的权重.

## 保存:

由于model变更为自行维护动态权重, 所以保存时也需要保存最新的权重, 以便于加载(测试)时使用.

**Comment by Tomorrowdawn** [comment]
![image](https://github.com/Tomorrowdawn/EvolveGCN/assets/61301081/f5562096-aa90-4c40-9f3a-f369cb622fd0)


如图所示, 训练循环发生在每(next)个step的多个batch上(这些batch可能不一样多).
